File: [UNIQUE] String/Programmers 2018 KAKAO BLIND RECRUITMENT.py
"""
문제 링크: https://school.programmers.co.kr/learn/courses/30/lessons/17686
문제 이름: [3차] 파일명 정렬
카테고리: 2018 KAKAO BLIND RECRUITMENT
최종 업로드 날짜: 2023.12.02
버전: v1
"""

"""
문제 풀이: state를 잘 분리해야 한다.
state 0은 head를 검사할 때, state 1은 number를 검사할 때, state 2는 tail을 검사할 때로 구분하였다.
또한, tail이 없이 number에서 끝나는 경우도 고려하였다.
"""
def split_file(files):
    split_arr = []
    for file_index, filename in enumerate(files):
        flag=0
        num_index=0
        head=""
        number=0
        tail=""
        for char_index, char in enumerate(filename):
            if char.isdigit()==True and flag==0: #처음으로 숫자를 마주친 경우
                head=filename[:char_index].lower()
                flag=1
                num_index=char_index
            if char_index==len(filename)-1 and flag==1: # 문자열의 마지막을 맞이한 경우
                number=int(filename[num_index:])
                tail=""
                break
            if char.isdigit()==False and flag==1: #숫자 다음으로 문자를 마주친 경우
                number=int(filename[num_index:char_index])
                tail=filename[char_index:]
                break
        split_tuple=(file_index, head, number, tail)
        print(split_tuple)
        split_arr.append(split_tuple)
        
    return split_arr

def custom_sort(item):
    return (item[1], item[2])

def answer_file(sorted_files, files):
    answer=[]
    for item in sorted_files:
        answer.append(files[item[0]]) # 원래 인덱스 값을 가져와서 순서대로 넣어주기
    return answer
            
def solution(files):
    split_arr = split_file(files)
    sorted_files = sorted(split_arr, key=custom_sort)
    answer = answer_file(sorted_files, files)
    return answer

File: [UNIQUE] Number/Programmers 2018 KAKAO BLIND RECRUITMENT.cpp
/*
문제 링크: https://school.programmers.co.kr/learn/courses/30/lessons/17687
문제 이름: [3차] n진수 게임
카테고리: 2018 KAKAO BLIND RECRUITMENT
최종 업로드 날짜: 2023.11.26
버전: v1
*/

/*
문제 풀이: translated를 통해서 튜브뿐만 아니라 모든 사람의 정답을 구해놓고,
translated 배열 안에서 다시 for 문을 돌면서 튜브의 정답을 찾는 방법을 선택했다.
그 이유는 하나의 숫자 안에서도 튜브가 선택해야 하는 정답이 여러 개 있을 수 있기 때문이다.
예시) "4"라는 숫자는 이진수로 "100"이므로, 맴버가 2명이고 튜브가 첫번째 순서라면 숫자 2개 선택
또한, 진수 변환하는 로직을 translate_number를 통해서 함수화하였다.
*/
#include <stdio.h>

#include <iostream>
#include <string>
#include <vector>

using namespace std;

string translate_number(int n, int present_num) {
    // 몫과 나머지 초기화
    int quotient = 100;
    int remainder = 100;
    string translated = "";

    while (quotient > 0) {
        quotient = present_num / n;
        remainder = present_num % n;
        string str_num = "";

        if (remainder >= 10 && remainder <= 15) {
            // 숫자->대문자 변환은 'A' 더하기
            char char_num = 'A' + (remainder - 10);
            str_num += char_num;
        } else {
            str_num = to_string(remainder);
        }

        translated = str_num + translated;
        // 다음 자리 수 구하기
        present_num = quotient;
    }
    return translated;
}

string solution(int n, int t, int m, int p) {
    string answer = "";
    string translated = "";

    // 진법 n, 미리 구할 숫자의 갯수 t, 게임에 참가하는 인원 m, 튜브의 순서 p
    // 구해야 하는 총 숫자 수(모든 맴버)=(t-1)*m+p
    int total_num = (t - 1) * m + p;
    int present_num = 0;
    while (1) {
        if (translated.length() >= total_num) {
            break;
        } else {
            translated += translate_number(n, present_num);
            present_num++;
        }
    }
    for (int i = p - 1; i <= total_num - 1; i += m) {
        // cout << i << " " << total_num << " " << translated[i] << "\n";
        answer += translated[i];
    }
    return answer;
}